### PR TITLE
GH-35409: [Python][Docs] Clarify S3FileSystem Credentials chain for EC2

### LINKED ISSUE #35409: [Docs] Document `S3FileSystem` behavior in EC2
### Context
While working with both `s3fs` and `pyarrow`, I noticed the difference in documentation regarding establishing AWS credentials. 

`s3fs` documentation explicitly mentions this scenario for using EC2
```
Boto will try the following methods, in order:
...
* for nodes on EC2, the IAM metadata provider
```
`pyarrow` documentation does not mention this behavior.

After digging into `pyarrow` code and the C++ implementation, I was able to find that `pyarrow` will use IAM metadata provide in EC2 to establish credentials. 

Here's the PR to explain in depth and the associated change to the `pyarrow.fs.S3FileSystem` documentation:

* MINOR: [S3FileSystem] Clarify S3FileSystem Credentials chain for EC2 #35312

### Component(s)

Documentation

### PR DESCRIPTION
**Pull request by kevinjqliu**
### Rationale for this change

When resolving AWS credentials on EC2 hosts, the underlying AWS SDK also looks at the EC2 Instance Metadata Service. 

I want to document this behavior for `pyarrow`.  The [`s3fs` documentation](https://s3fs.readthedocs.io/en/latest/#credentials) mention this specific case for EC2.

### What changes are included in this PR?

Documentation for the behavior described above. 

#### Technical Details
`S3FileSystem` uses the [`CS3Options.Defaults()`](https://github.com/apache/arrow/blob/5de56928e0fe43f02005552eee058de57ffb2682/python/pyarrow/_s3fs.pyx#L317) option when no credentials are passed into the constructor.  It utilizes the [`Aws::Auth::DefaultAWSCredentialsProviderChain`](https://github.com/apache/arrow/blob/1de159d0f6763766c19b183dd309b8757723b43a/cpp/src/arrow/filesystem/s3fs.cc#L213)

The C++ implementation of [`DefaultAWSCredentialsProviderChain`](https://sdk.amazonaws.com/cpp/api/0.14.3/class_aws_1_1_auth_1_1_default_a_w_s_credentials_provider_chain.html) not only [reads the environment variable](https://sdk.amazonaws.com/cpp/api/0.14.3/class_aws_1_1_auth_1_1_environment_a_w_s_credentials_provider.html) when trying to resolve AWS credentials, but also [looks at profile config](https://sdk.amazonaws.com/cpp/api/0.14.3/class_aws_1_1_auth_1_1_profile_config_file_a_w_s_credentials_provider.html) and the [EC2 Instance Metadata Service](https://sdk.amazonaws.com/cpp/api/0.14.3/class_aws_1_1_auth_1_1_instance_profile_credentials_provider.html). 

### Are these changes tested?

No, just documentation changes

### Are there any user-facing changes?

Yes, changing public documentation

* Closes: #35409

### Render Changes
Render the changes locally via [Building the doc](https://arrow.apache.org/docs/developers/documentation.html#building-docs): 
`docs/source/python/filesystems.rst`:
Screenshot 2023-07-30 at 6 22 02 PM

`python/pyarrow/_s3fs.pyx`:
Screenshot 2023-07-31 at 3 31 30 PM

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
@kevinjqliu thanks for the contribution!

We have a section about S3FileSystem in the user guide as well: https://arrow.apache.org/docs/dev/python/filesystems.html#s3. Could you have a look to see if that might require an equivalent update?

File: python/pyarrow/_s3fs.pyx [hunk 1]
```diff
@@ -140,13 +140,17 @@ cdef class S3FileSystem(FileSystem):
...
-    otherwise both access_key and secret_key must be provided.
-
+    AWS access_key and secret_key can be provided explicitly.
+    
     If role_arn is provided instead of access_key and secret_key, temporary
     credentials will be fetched by issuing a request to STS to assume the
     specified role.
+    
+    If neither access_key nor secret_key are provided, and role_arn is also not
+    provided, then attempts to initialize from the Default AWS Credentials Provider Chain in order of:
```
Comment (jorisvandenbossche) [inline python/pyarrow/_s3fs.pyx:150]: Formatting nit: can you limit the line length here?

File: python/pyarrow/_s3fs.pyx [hunk 2]
```diff
@@ -140,13 +140,17 @@ cdef class S3FileSystem(FileSystem):
...
+    * `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and `AWS_SESSION_TOKEN` environment variables
```
Comment (jorisvandenbossche) [inline python/pyarrow/_s3fs.pyx:151]:
```suggestion

    * `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and `AWS_SESSION_TOKEN` environment variables
```

RestructuredText markup needs a blank line before a bullet point list

**Comment by kevinjqliu** [comment]
@jorisvandenbossche fixed a formatting issue. there was something wrong with CI too, hence the force pushes.

**Comment by pitrou** [comment]
@jorisvandenbossche Would you like to take another look?

**Comment by kevinjqliu** [comment]
hey, @jorisvandenbossche wanted to ping on this again. thanks!

**Comment by kou** [comment]
@github-actions crossbow submit docs-preview

File: python/pyarrow/_s3fs.pyx [hunk 3]
```diff
@@ -140,14 +140,20 @@ cdef class S3FileSystem(FileSystem):
...
     If role_arn is provided instead of access_key and secret_key, temporary
     credentials will be fetched by issuing a request to STS to assume the
     specified role.
 
+    If neither access_key nor secret_key are provided, and role_arn is also not
+    provided, then attempts establish the credentials automatically.
+    S3FileSystem will try the following methods, in order:
+
+    * `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and `AWS_SESSION_TOKEN` environment variables
+    * configuration files such as `~/.aws/credentials` and `~/.aws/config`
```
Comment (kou) [inline python/pyarrow/_s3fs.pyx:154]: We need two `` ` ``s for reStructuredText:

```suggestion
    * ``AWS_ACCESS_KEY_ID``, ``AWS_SECRET_ACCESS_KEY``, and ``AWS_SESSION_TOKEN`` environment variables
    * configuration files such as ``~/.aws/credentials`` and ``~/.aws/config``
```

File: python/pyarrow/_s3fs.pyx [hunk 4]
```diff
@@ -140,14 +140,20 @@ cdef class S3FileSystem(FileSystem):
...
-    provided, then attempts to initialize from AWS environment variables,
-    otherwise both access_key and secret_key must be provided.
+    AWS access_key and secret_key can be provided explicitly.
 
     If role_arn is provided instead of access_key and secret_key, temporary
     credentials will be fetched by issuing a request to STS to assume the
     specified role.
 
+    If neither access_key nor secret_key are provided, and role_arn is also not
+    provided, then attempts establish the credentials automatically.
```
Comment (kou) [inline python/pyarrow/_s3fs.pyx:150]: Is `to` missing?

```suggestion
    provided, then attempts to establish the credentials automatically.
```

**Comment by kou** [comment]
@github-actions crossbow submit preview-docs

**Comment by kevinjqliu** [comment]
thanks @kou I've applied your suggestions

**Comment by kou** [comment]
@github-actions crossbow submit preview-docs

**Comment by kevinjqliu** [comment]
@kou built the changes locally, I've attached a screenshot above

**Comment by kou** [comment]
Thanks but we have 2 changed pages but the screenshot shows only 1 page.

**Comment by kevinjqliu** [comment]
The visible change is from `docs/source/python/filesystems.rst`. 
`python/pyarrow/_s3fs.pyx` is a code comment change.

**Comment by kou** [comment]
We use it for API reference: https://arrow.apache.org/docs/python/generated/pyarrow.fs.S3FileSystem.html#pyarrow.fs.S3FileSystem

**Comment by kevinjqliu** [comment]
> We use it for API reference: https://arrow.apache.org/docs/python/generated/pyarrow.fs.S3FileSystem.html#pyarrow.fs.S3FileSystem

I see. I'm following [Building a single directory for dev purposes without all the pre-requisites](https://arrow.apache.org/docs/developers/documentation.html#building-a-single-directory-for-dev-purposes-without-all-the-pre-requisites) to render the changes for `docs/source/python`.

The changes in `python/pyarrow/_s3fs.pyx` does not show up. How can I render that part?

**Comment by kou** [comment]
We need pyarrow to build documents in `python/pyarrow/_s3fs.pyx`. So we can't use the approach.

We need to use the full (at least including pyarrow) build approach.
https://arrow.apache.org/docs/developers/documentation.html#building-with-docker is easier than https://arrow.apache.org/docs/developers/documentation.html#building .

**Comment by kou** [comment]
Could you rebase on main to use #36943 ?

**Comment by kevinjqliu** [comment]
@kou I've rebased the branch and was able to build it locally using docker. Updated the description above to render changes to both files. 

Thanks for guiding me through the process

**Comment by kou** [comment]
@github-actions crossbow submit preview-docs

**Comment by kou** [comment]
http://crossbow.voltrondata.com/pr_docs/35312/

* http://crossbow.voltrondata.com/pr_docs/35312/python/filesystems.html#s3
* http://crossbow.voltrondata.com/pr_docs/35312/python/generated/pyarrow.fs.S3FileSystem.html

**Review by kou — APPROVED** [review]
+1

**Pull request merged**